src/ppx_import.ml

```diff
@@ -182,311 +164,414 @@ let get_modtype_decl ~loc sig_items parent_lid elem =
+               let open Ocaml_common in
+               let label = Location.mknoloc label in
+               let desc =
+                 match row_field with
+                 | Types.Rpresent None -> Rtag (label, true, [])
+                 | Types.Rpresent (Some ttyp) ->
+                     Rtag (label, false, [ core_type_of_type_expr ~subst ttyp ])
+                 | _ -> assert false
+               in
+               { prf_desc = desc; prf_loc = !default_loc; prf_attributes = [] })
+      in
+      Typ.variant fields Closed None
+  | _ -> assert false
 
-let ptype_decl_of_ttype_decl ~manifest ~subst ptype_name ttype_decl =
+let ptype_decl_of_ttype_decl ~manifest ~subst ptype_name
+    (ttype_decl : Ocaml_common.Types.type_declaration) : Ppxlib.type_declaration
+    =
+  let open Ppxlib in
   let subst =
     match manifest with
-    | Some { ptyp_desc = Ptyp_constr (_, ptype_args); ptyp_loc ; _ } ->
-      subst @ begin try
-          List.map2 (fun tparam pparam ->
-            match tparam with
-              | { desc = Tvar (Some var) ; _ } -> [`Var var, pparam]
-              | { desc = Tvar None ; _ }       -> []
+    | Some { ptyp_desc = Ptyp_constr (_, ptype_args); ptyp_loc; _ } -> (
+        subst
+        @
+        try
+          List.map2
+            (fun (tparam : Ocaml_common.Types.type_expr) pparam ->
+              match tparam with
+              | { desc = Tvar (Some var); _ } -> [ (`Var var, pparam) ]
+              | { desc = Tvar None; _ } -> []
               | _ -> assert false)
             ttype_decl.type_params ptype_args
-        |> List.concat
-      with Invalid_argument _ ->
-        raise_errorf ~loc:ptyp_loc "Imported type has %d parameter(s), but %d are passed"
-                                   (List.length ttype_decl.type_params)
-                                   (List.length ptype_args)
-      end
+          |> List.concat
+        with Invalid_argument _ ->
+          Location.raise_errorf ~loc:ptyp_loc
+            "Imported type has %d parameter(s), but %d are passed"
+            (List.length ttype_decl.type_params)
+            (List.length ptype_args))
     | None -> []
     | _ -> assert false
   in
   let ptype_params =
-    List.map2 (fun param _variance ->
-        core_type_of_type_expr ~subst param,
-        (* The equivalent of not specifying the variance explicitly.
-           Since the very purpose of ppx_import is to include the full definition,
-           it should always be sufficient to rely on the inferencer to deduce variance. *)
-        Invariant)
+    List.map2
+      (fun param _variance ->
+        ( core_type_of_type_expr ~subst param,
+          (* The equivalent of not specifying the variance explicitly.
+             Since the very purpose of ppx_import is to include the full definition,
+             it should always be sufficient to rely on the inferencer to deduce variance. *)
+          (Ppxlib.Asttypes.NoVariance, Ppxlib.Asttypes.NoInjectivity) ))
```

Comment:
Definitely not sure about that one. Is `Ppxlib.Asttypes.NoInjectivity` correct ?

Reply:
This is indeed tricky stuff, @gasche do you know who could help with this?

Reply:
I just looked at this (sorry for the delay). It's a bit tricky but I think that the current code in the PR is correct:
- for abstract type, the default is "non injective", so `NoInjectivity` is the conservative choice
- non-abstract types are always injective, so you could use `Injective` here, but it looks like the OCaml parser uses `NoInjective` by default as well and lets the type-checker figure things out afterwards.